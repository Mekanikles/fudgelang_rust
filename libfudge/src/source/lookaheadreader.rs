use crate::source::Source;

pub struct LookAheadSourceReader<'a> {
    source: &'a Source,
    pos: u64,
    current: Option<u8>,
    lookahead: Option<u8>,
}

impl<'a> LookAheadSourceReader<'a> {
    pub fn new(source: &'a Source) -> Self {
        let data = source.data();
        let current = data.get(0);
        let lookahead = current.and_then(|_| data.get(1));
        Self {
            source,
            pos: 0,
            current: current.cloned(),
            lookahead: lookahead.cloned(),
        }
    }
    pub fn advance(&mut self) {
        if self.current == None {
            return;
        }
        self.pos += 1;
        self.current = self.lookahead;
        self.lookahead = self.source.data().get(self.pos as usize + 1).cloned();
    }
    pub fn pos(&self) -> u64 {
        self.pos
    }
    pub fn peek(&self) -> Option<u8> {
        self.current
    }
    pub fn lookahead(&self) -> Option<u8> {
        self.lookahead
    }

    pub fn source(&self) -> &'a Source {
        return self.source;
    }
}
