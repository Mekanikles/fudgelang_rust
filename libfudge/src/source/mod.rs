use std::fs;
use std::path::*;

mod lookaheadreader;
pub use lookaheadreader::*;

use std::io::BufRead;

#[derive(Copy, Clone, PartialEq, Debug)]
pub struct SourceSpan {
    pub pos: u64,
    pub len: usize,
}

pub struct LineInfo {
    pub text: String,
    pub line_start: usize,
    pub row: u32,
}

pub struct Source {
    name: String,
    data: Vec<u8>,
}

impl Source {
    pub fn from_file<P: AsRef<Path>>(file: P) -> Source {
        Source {
            name: file
                .as_ref()
                .file_name()
                .unwrap()
                .to_str()
                .unwrap()
                .to_string(),
            data: fs::read(file).unwrap(),
        }
    }

    pub fn from_str(data: &str) -> Source {
        Source {
            name: "stringsource".into(),
            data: data.as_bytes().to_vec(),
        }
    }

    pub fn from_bytes(bytes: &[u8]) -> Source {
        Source {
            name: "bytesource".into(),
            data: bytes.to_vec(),
        }
    }

    pub fn name(&self) -> &str {
        return self.name.as_str();
    }

    pub fn data(&self) -> &[u8] {
        return &self.data[..];
    }

    pub fn to_str(&self) -> &str {
        return std::str::from_utf8(&self.data[..]).unwrap();
    }

    pub fn get_span(&self, span: &SourceSpan) -> &[u8] {
        return &self.data[span.pos as usize..(span.pos as usize + span.len)];
    }

    pub fn get_source_string(&self, span: &SourceSpan) -> &str {
        return std::str::from_utf8(self.get_span(&span)).unwrap();
    }

    pub fn get_line_info(&self, filepos: u64) -> Option<LineInfo> {
        let mut seekpos = 0;
        let mut row = 0;
        let mut text = String::new();

        while let Ok(bytes_read) = (&self.data[seekpos..]).read_line(&mut text) {
            if bytes_read == 0 {
                return None;
            }

            let eol = seekpos + bytes_read;
            if eol > filepos as usize {
                return Some(LineInfo {
                    text,
                    line_start: seekpos,
                    row: row + 1,
                });
            }
            seekpos = eol;
            row += 1;
            text.clear();
        }
        return None;
    }
}

#[cfg(test)]
mod test;
