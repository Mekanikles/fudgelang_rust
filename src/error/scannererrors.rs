use super::*;

pub fn new_unexpected_eof_error(pos: u64, message: String) -> Error {
    Error::at_span(
        errors::UnexpetedEOF,
        source::SourceSpan {
            pos,
            len: 1,
        },
        message,
    )
}

pub fn new_non_utf8_sequence_error(pos: u64, len: u64) -> Error {
    Error::at_span(
        errors::NonUtf8Sequence,
        source::SourceSpan {
            pos,
            len: len as usize,
        },
        "Found non-ut8 sequence".into(),
    )
}

pub fn new_invalid_sequence_error(pos: u64, len: u64) -> Error {
    Error::at_span(
        errors::InvalidSequece,
        source::SourceSpan {
            pos,
            len: len as usize,
        },
        "Found invalid sequence".into(),
    )
}

pub fn new_unexpected_sequence_error(pos: u64, len: u64, message: String) -> Error {
    Error::at_span(
        errors::UnexpectedSequence,
        source::SourceSpan {
            pos,
            len: len as usize,
        },
        message,
    )
}

pub fn new_non_ascii_identifier_error(pos: u64, len: u64, identifier: String) -> Error {
    Error::at_span(
        errors::NonAsciiIdentifier,
        source::SourceSpan {
            pos,
            len: len as usize,
        },
        format!("Non-ascii identifier: '{}'", identifier),
    )
}

pub fn new_invalid_indentation_error(pos: u64, len: u64) -> Error {
    Error::at_span(
        errors::InvalidIndentation,
        source::SourceSpan {
            pos,
            len: len as usize,
        },
        "Indentations are only allowed at the start of a line or immediately after other indentations".into(),
    )
}
