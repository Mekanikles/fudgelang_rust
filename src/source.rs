use std::io::Read;
use std::io::BufReader;
use std::option::Option;

pub trait SourceReader {
    fn advance(&mut self);
    fn pos(&self) -> u64;
    fn peek(&self) -> Option<u8>;
}

pub struct BufferedSourceReader<R> {
    reader : BufReader<R>,
    pos : u64,
    current : Option<u8>,
}

impl<R : Read> BufferedSourceReader<R> {
    pub fn new(read : R) -> BufferedSourceReader<R> {
        let mut reader = BufferedSourceReader { 
            reader: BufReader::new(read), 
            pos: 0,
            current: None };
        reader.advance();
        return reader;
    }
}

impl<R : Read> SourceReader for BufferedSourceReader<R> {
    fn advance(&mut self) {
        let mut buf = [0; 1];
        self.current = match self.reader.read(&mut buf) {
            Ok(n) if n > 0 => Some(buf[0]),
            _ => None };
    }
    fn pos(&self) -> u64 { self.pos }
    fn peek(&self) -> Option<u8> { self.current }
}

pub trait Source {
    type SourceReaderT : SourceReader;
    fn get_reader(&self) -> Self::SourceReaderT;
}

pub struct ByteArraySource<'a> {
    bytes : &'a[u8],
}

impl<'a> ByteArraySource<'a> {
    pub fn new(bytes : &'a[u8]) -> ByteArraySource { ByteArraySource { bytes } }
}

impl<'a> Source for ByteArraySource<'a> {
    type SourceReaderT = BufferedSourceReader<&'a[u8]>;
    fn get_reader(&self) -> Self::SourceReaderT { 
        return BufferedSourceReader::new(self.bytes); 
    }
}