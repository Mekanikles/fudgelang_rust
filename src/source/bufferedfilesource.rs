use super::*;
use std::fs::File;
use std::io::BufReader;
use std::path::*;

pub struct BufferedFileSource {
    filename: PathBuf,
}

impl BufferedFileSource {
    pub fn from_filepath<P: AsRef<Path>>(filename: P) -> BufferedFileSource {
        BufferedFileSource {
            filename: PathBuf::from(filename.as_ref()),
        }
    }
}

impl<'a> Source<'a, BufReader<File>> for BufferedFileSource {
    fn get_name(&'a self) -> &'a str {
        return self.filename.to_str().unwrap();
    }

    fn get_readable(&self) -> BufReader<File> {
        BufReader::new(File::open(self.filename.clone()).expect("File not found!"))
    }
}
