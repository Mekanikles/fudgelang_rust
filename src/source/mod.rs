use std::io::Read;
use std::io::BufReader;
use std::option::Option;

mod filesource; pub use filesource::*;
mod memorysource; pub use memorysource::*;

#[cfg(test)]
mod test;

pub struct SourceReader<R> {
    reader : BufReader<R>,
    pos : u64,
    current : Option<u8>,
    lookahead : Option<u8>,
}

impl<R : Read> SourceReader<R> {
    fn readbyte(reader : &mut BufReader<R>) -> Option<u8>
    {
        let mut buf = [0; 1];
        match reader.read(&mut buf) {
            Ok(n) if n > 0 => {
                Some(buf[0])},
            _ => None
        }
    }
    pub fn new(read : R) -> SourceReader<R> {
        let mut reader = BufReader::new(read);
        let mut pos = 0;
        let mut lookahead = None;
        let current = Self::readbyte(&mut reader);
        if current != None {
            pos = 1;
            lookahead = Self::readbyte(&mut reader);
        }

        return SourceReader { 
            reader, 
            pos,
            current,
            lookahead };
    }
    pub fn advance(&mut self) {
        if self.current == None {
            return;
        }
        self.pos += 1;
        self.current = self.lookahead;
        self.lookahead = Self::readbyte(&mut self.reader);
    }
    pub fn pos(&self) -> u64 { self.pos }
    pub fn peek(&self) -> Option<u8> { self.current }
    pub fn lookahead(&self) -> Option<u8> { self.lookahead } 
}

pub trait Source<'a, R> {
    fn get_reader(&'a self) -> SourceReader<R>;
}


