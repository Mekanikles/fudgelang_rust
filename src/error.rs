use crate::scanner::token;
use crate::source;

#[derive(Clone, Debug, PartialEq)]
pub enum ErrorId {
    NonAsciiChar,
    UnexpectedChar,
    NonUtf8Sequence,
    EnexpectedSequence,
}

#[derive(Clone)]
pub struct Error {
    pub id: ErrorId,
    pub message: String,
    pub source_span: source::SourceSpan,
}

impl Error {
    pub fn at_span(id: ErrorId, source_span: source::SourceSpan, message: String) -> Error {
        Error {
            id,
            message,
            source_span,
        }
    }
    pub fn at_pos(id: ErrorId, pos: u64, message: String) -> Error {
        Self::at_span(id, source::SourceSpan { pos, len: 1 }, message)
    }
    pub fn at_token(id: ErrorId, token: &token::Token, message: String) -> Error {
        Self::at_span(id, token.source_span, message)
    }
}

pub fn new_non_utf8_sequence_error(pos: u64, len: u64) -> Error {
    Error::at_span(
        ErrorId::NonUtf8Sequence,
        source::SourceSpan {
            pos,
            len: len as usize,
        },
        "Found non-ut8 sequence".into(),
    )
}

pub fn new_unexpected_char_error(c: char, pos: u64) -> Error {
    Error::at_pos(
        ErrorId::UnexpectedChar,
        pos,
        format!("Found unexpected character '{}'", c),
    )
}

pub fn new_non_ascii_char_error(c: char, pos: u64) -> Error {
    Error::at_pos(
        ErrorId::NonAsciiChar,
        pos,
        format!(
            "Found non-ascii character '{}' outside of string literal",
            c
        ),
    )
}

pub fn new_unexpected_sequence_error(pos: u64, len: u64, message: String) -> Error {
    Error::at_span(
        ErrorId::EnexpectedSequence,
        source::SourceSpan {
            pos,
            len: len as usize,
        },
        message,
    )
}
