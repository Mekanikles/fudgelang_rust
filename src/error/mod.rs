use crate::scanner::token;
use crate::source;

pub mod scannererrors;
pub use scannererrors::*;

pub mod errors {
    // TODO: We might want to attribute "severity" to errors more dynamically
    //  (say, via some map), so that users can elevate warnings to errors etc
    // The static division here might be better oriented towards recoverability, 
    //  or category, rather than user-facing severity
    pub use FatalErrorType::*;
    pub use MajorErrorType::*;
    pub use MinorErrorType::*;

    #[derive(Clone, Copy, Debug, PartialEq)]
    pub enum FatalErrorType {
        UnexpectedEOF,
        ErrorLimitExceeded,
    }

    #[derive(Clone, Copy, Debug, PartialEq)]
    pub enum MajorErrorType {
        InvalidSequece,
        NonUtf8Sequence,
        UnexpectedSequence,
        UnexpectedToken,
        ExpectedExpression,
        UnknownCompilerDirective,
    }

    #[derive(Clone, Copy, Debug, PartialEq)]
    pub enum MinorErrorType {
        NonAsciiIdentifier,
        InvalidIndentation,
        ExpectedInputParameterDeclaration,
        ExpectedOutputParameterDeclaration,
    }
}

const FATAL_ERROR_THRESHOLD: usize = 1;
const MAJOR_ERROR_THRESHOLD: usize = 5;
const MINOR_ERROR_THRESHOLD: usize = 20;

#[derive(Clone, Copy, Debug, PartialEq)]
pub enum ErrorId {
    FatalError(errors::FatalErrorType),
    MajorError(errors::MajorErrorType),
    MinorError(errors::MinorErrorType),
}

pub trait ErrorIdConstructor {
    fn create_id(&self) -> ErrorId;
}

impl ErrorIdConstructor for errors::FatalErrorType {
    fn create_id(&self) -> ErrorId {
        ErrorId::FatalError(*self)
    }
}

impl ErrorIdConstructor for errors::MajorErrorType {
    fn create_id(&self) -> ErrorId {
        ErrorId::MajorError(*self)
    }
}

impl ErrorIdConstructor for errors::MinorErrorType {
    fn create_id(&self) -> ErrorId {
        ErrorId::MinorError(*self)
    }
}

pub fn new_error_id<T: ErrorIdConstructor>(t: T) -> ErrorId {
    t.create_id()
}

pub fn error_label(id: ErrorId) -> &'static str {
    match id {
        ErrorId::FatalError(_e) => {
            return "Error";
        }
        ErrorId::MajorError(_e) => {
            return "Error";
        }
        ErrorId::MinorError(_e) => {
            return "Error";
        }
    }
}

pub fn error_code(id: ErrorId) -> String {
    match id {
        ErrorId::FatalError(e) => {
            return format!("A{:03}", e as i32);
        }
        ErrorId::MajorError(e) => {
            return format!("B{:03}", e as i32);
        }
        ErrorId::MinorError(e) => {
            return format!("C{:03}", e as i32);
        }
    }
}

#[derive(Clone)]
pub struct Error {
    pub id: ErrorId,
    pub message: String,
    pub source_span: source::SourceSpan,
}

impl Error {
    pub fn at_span<T: ErrorIdConstructor>(
        t: T,
        source_span: source::SourceSpan,
        message: String,
    ) -> Error {
        Error {
            id: new_error_id(t),
            message,
            source_span,
        }
    }
    pub fn at_pos<T: ErrorIdConstructor>(t: T, pos: u64, message: String) -> Error {
        Self::at_span(t, source::SourceSpan { pos, len: 1 }, message)
    }
    pub fn at_token<T: ErrorIdConstructor>(t: T, token: &token::Token, message: String) -> Error {
        Self::at_span(t, token.source_span, message)
    }
}

pub struct ErrorData {
    fatal_error_count: usize,
    major_error_count: usize,
    minor_error_count: usize,
    errors: Vec<Error>,
}

pub struct ErrorManager {
    reached_error_limit : bool,
    error_data : ErrorData,
}

impl ErrorManager {
    pub fn new() -> Self {
        ErrorManager {
            reached_error_limit: false,
            error_data: ErrorData {
                fatal_error_count: 0,
                major_error_count: 0,
                minor_error_count: 0,
                errors: Vec::new(),
            },
        }
    }
    
    pub fn reached_error_limit(&self) -> bool {
        return self.reached_error_limit;
    }

    pub fn get_errors(&self) -> &Vec<Error> {
        return &self.error_data.errors;
    }

    pub fn log_error(&mut self, error: Error) -> ErrorId {
        let id = error.id;
        self.error_data.errors.push(error);

        match id {
            ErrorId::FatalError(_e) => {
                self.error_data.fatal_error_count += 1;
                if self.error_data.fatal_error_count >= FATAL_ERROR_THRESHOLD {
                    self.reached_error_limit = true;
                }
            }
            ErrorId::MajorError(_e) => {
                self.error_data.major_error_count += 1;
                if self.error_data.major_error_count >= MAJOR_ERROR_THRESHOLD {
                    self.reached_error_limit = true;
                }
            }
            ErrorId::MinorError(_e) => {
                self.error_data.minor_error_count += 1;
                if self.error_data.minor_error_count >= MINOR_ERROR_THRESHOLD {
                    self.reached_error_limit = true;
                }
            }
        }
        return id;
    }

    pub fn adjust_last_error_end(&mut self, end : u64) {
        let pos = self.error_data.errors.last_mut().unwrap().source_span.pos;
        self.error_data.errors.last_mut().unwrap().source_span.len = (end - pos) as usize;
    }
}