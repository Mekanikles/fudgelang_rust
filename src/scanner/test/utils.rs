use crate::error;
use crate::source::*;
use super::*;

pub fn expect_error_ids(errors: &Vec<error::Error>, expected_error_ids: &[error::ErrorId]) {
    for i in 0..expected_error_ids.len() {
        assert_eq!(errors[i].id, expected_error_ids[i]);
    }
    assert_eq!(
        errors.len(),
        expected_error_ids.len(),
        "Found more errors than expected!"
    );
}

pub fn expect_token(expected_tokens: &[Token], i: usize, scanned_token: &Token) {
    if i < expected_tokens.len() {
        assert_eq!(expected_tokens[i], *scanned_token);
    } else {
        panic!("Scanned more tokens than expected!");
    }
}

pub fn verify_exact_scanner_tokens<S: Scanner>(scanner: &mut S, expected_tokens: &[Token]) {
    let mut count = 0;
    while let Some(t) = scanner.read_token() {
        expect_token(expected_tokens, count, &t);
        count += 1;
    }

    assert_eq!(
        count as usize,
        expected_tokens.len(),
        "Scanned less tokens than expected!"
    );
}

pub fn verify_sparse_scanner_tokens<S: Scanner>(scanner: &mut S, expected_tokens: &[Token]) {
    let mut scanned_tokens = Vec::new();
    while let Some(t) = scanner.read_token() {
        scanned_tokens.push(t);
    }

    for t in expected_tokens {
        assert!(
            scanned_tokens.iter().position(|e| e == t) != None,
            "Expected token not found in scan!"
        );
    }
}

// Checks that the scanner produces an exact list of tokens
pub fn verify_exact_scan(source: &str, expected_tokens: &[Token]) -> Vec<error::Error> {
    let source = MemorySource::from_str(source);
    let mut scanner = ScannerImpl::new(&source);

    verify_exact_scanner_tokens(&mut scanner, expected_tokens);
    scanner.errors.clone()
}

// Checks that scanner produces any tokens that matches the list
pub fn verify_sparse_scan(source: &str, expected_tokens: &[Token]) -> Vec<error::Error> {
    let source = MemorySource::from_str(source);
    let mut scanner = ScannerImpl::new(&source);

    verify_sparse_scanner_tokens(&mut scanner, expected_tokens);
    scanner.errors.clone()
}

pub fn do_scan(source: &str) -> Vec<error::Error> {
    let source = MemorySource::from_str(source);
    let mut scanner = ScannerImpl::new(&source);
    while scanner.read_token().is_some() {};
    scanner.errors.clone()
}
