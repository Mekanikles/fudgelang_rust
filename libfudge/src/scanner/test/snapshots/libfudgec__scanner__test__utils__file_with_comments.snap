---
source: src/scanner/test/utils.rs
expression: scanned_tokens

---
[
    Token {
        tokentype: Identifier,
        source_span: SourceSpan {
            pos: 0,
            len: 5,
        },
    },
    Token {
        tokentype: Comma,
        source_span: SourceSpan {
            pos: 5,
            len: 1,
        },
    },
    Token {
        tokentype: Identifier,
        source_span: SourceSpan {
            pos: 7,
            len: 7,
        },
    },
    Token {
        tokentype: Dot,
        source_span: SourceSpan {
            pos: 14,
            len: 1,
        },
    },
    Token {
        tokentype: Comment,
        source_span: SourceSpan {
            pos: 16,
            len: 10,
        },
    },
    Token {
        tokentype: LineBreak,
        source_span: SourceSpan {
            pos: 26,
            len: 1,
        },
    },
    Token {
        tokentype: Identifier,
        source_span: SourceSpan {
            pos: 27,
            len: 5,
        },
    },
    Token {
        tokentype: Comment,
        source_span: SourceSpan {
            pos: 32,
            len: 18,
        },
    },
    Token {
        tokentype: Identifier,
        source_span: SourceSpan {
            pos: 50,
            len: 5,
        },
    },
    Token {
        tokentype: Comma,
        source_span: SourceSpan {
            pos: 55,
            len: 1,
        },
    },
    Token {
        tokentype: LineBreak,
        source_span: SourceSpan {
            pos: 56,
            len: 1,
        },
    },
    Token {
        tokentype: LineBreak,
        source_span: SourceSpan {
            pos: 57,
            len: 1,
        },
    },
    Token {
        tokentype: Comment,
        source_span: SourceSpan {
            pos: 58,
            len: 94,
        },
    },
    Token {
        tokentype: LineBreak,
        source_span: SourceSpan {
            pos: 152,
            len: 1,
        },
    },
    Token {
        tokentype: Identifier,
        source_span: SourceSpan {
            pos: 153,
            len: 7,
        },
    },
    Token {
        tokentype: Comma,
        source_span: SourceSpan {
            pos: 160,
            len: 1,
        },
    },
    Token {
        tokentype: Identifier,
        source_span: SourceSpan {
            pos: 164,
            len: 5,
        },
    },
    Token {
        tokentype: Dot,
        source_span: SourceSpan {
            pos: 169,
            len: 1,
        },
    },
    Token {
        tokentype: LineBreak,
        source_span: SourceSpan {
            pos: 170,
            len: 1,
        },
    },
    Token {
        tokentype: Comment,
        source_span: SourceSpan {
            pos: 171,
            len: 16,
        },
    },
]
